vars/yamlSubstituteVariables.groovy

```diff
@@ -0,0 +1,244 @@
+
+        script.commonPipelineEnvironment.setValue('yamlSubstituteVariablesResult', result)
+    }
+}
+
+/**
+ * Recursively substitutes all variables inside the object tree of the manifest YAML.
+ * @param manifestNode - the manifest YAML to replace variables in.
+ * @param variablesData - the variables values.
+ * @param context - an execution context that can be used to query if any variables were replaced.
+ * @return a YAML object graph which has all variables replaced.
+ */
+private Object substitute(Object manifestNode, Object variablesData, ExecutionContext context) {
+    Map<String, Object> variableSubstitutes = getVariableSubstitutes(variablesData)
+
+    if (containsVariableReferences(manifestNode)) {
+
+        Object complexResult = null
+        String stringNode = manifestNode as String
+        Map<String, String> referencedVariables = getReferencedVariables(stringNode)
+        referencedVariables.each { referencedVariable ->
+            String referenceToReplace = referencedVariable.getKey()
+            String referenceName = referencedVariable.getValue()
+            Object substitute = variableSubstitutes.get(referenceName)
+
+            if (null == substitute) {
+                echo  "[YamlSubstituteVariables] ERROR - Found variable reference ${referenceToReplace} in input Yaml but no variable value to replace it with."
+                echo  "[YamlSubstituteVariables] ERROR - Leaving it unresolved. Check your variables Yaml data and make sure the variable is properly declared."
+                echo  "[YamlSubstituteVariables] ERROR - Unresolved variables may lead to follow-up problems (e.g. during a CF deployment). Failing this build."
+                error "[YamlSubstituteVariables] Not all variables could be resolved."
+            }
+
+            echo "Replacing: ${referenceToReplace} with ${substitute}"
+
+            if(isSingleVariableReference(stringNode)) {
+                logger.debug("Node ${stringNode} is SINGLE variable reference. Substitute type is: ${substitute.getClass().getName()}")
+                // if the string node we need to do replacements for is
+                // a reference to a single variable, i.e. should be replaced
+                // entirely with the variable value, we replace the entire node
+                // with the variable's value (which can possibly be a complex type).
+                complexResult = substitute
+            }
+            else {
+                logger.debug("Node ${stringNode} is multi-variable reference or contains additional string constants. Substitute type is: ${substitute.getClass().getName()}")
+                // if the string node we need to do replacements for contains various
+                // variable references or a variable reference and constant string additions
+                // we do a string replacement of the variables inside the node.
+                String regex = "\\(\\(${referenceName}\\)\\)"
```

Comment:
Is the `((variable))` syntax (I mean the double `(` / `)` some kind of a (de facto) standard for yaml? What comes into my mind first when thinking about placeholders / variables is the well known `${foo}` syntax.

Reply:
Actually, I only found variable references from *within* the same Yaml file standardized.
The syntax used here is the syntax [used by Cloud Foundry](https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html#variable-substitution).

So I guess you could say, it is a de-facto standard. Aditionally, `{`and `}` are special characters in Yaml and need to be escaped / used only in quotes [[ref](https://stackoverflow.com/questions/19109912/do-i-need-quotes-for-strings-in-yaml)].

The format and link to CF specs is part of the API documentation of `YamlUtils` and `cfManifestSubstituteVariables`.